libobs/obs-encoder.c

```diff
@@ -184,6 +184,110 @@ static inline bool gpu_encode_available(const struct obs_encoder *encoder)
 	       (video->using_p010_tex || video->using_nv12_tex);
 }
 
+static void maybe_set_gpu_video_media(struct obs_encoder *encoder)
+{
+	struct obs_core_video_mix *mix = NULL;
+	bool create_mix = true;
+	struct obs_video_info ovi;
+	const struct video_output_info *info;
+
+	if (!encoder->media)
+		return;
+
+	info = video_output_get_info(encoder->media);
+
+	if (!encoder->gpu_scaling_enabled)
+		return;
+
+	if (!encoder->scaled_height && !encoder->scaled_width)
+		return;
+
+	pthread_mutex_lock(&obs->video.mixes_mutex);
+	for (size_t i = 0; i < obs->video.mixes.num; i++) {
+		struct obs_core_video_mix *current = obs->video.mixes.array[i];
+		const struct video_output_info *voi =
+			video_output_get_info(current->video);
+		if (current->view != &obs->data.main_view)
+			continue;
+
+		if (voi->width != encoder->scaled_width ||
+		    voi->height != encoder->scaled_height)
+			continue;
+
+		if (voi->format != info->format ||
+		    voi->colorspace != info->colorspace ||
+		    voi->range != info->range)
+			continue;
+
+		current->encoder_refs += 1;
+		obs_encoder_set_video(encoder, current->video);
+		create_mix = false;
+		break;
+	}
+
+	if (!obs->video.main_mix) {
+		create_mix = false;
+	} else {
+		ovi = obs->video.main_mix->ovi;
+	}
+
+	pthread_mutex_unlock(&obs->video.mixes_mutex);
+
+	if (!create_mix)
+		return;
+
+	ovi.output_format = info->format;
+	ovi.colorspace = info->colorspace;
+	ovi.range = info->range;
+
+	ovi.output_height = encoder->scaled_height;
+	ovi.output_width = encoder->scaled_width;
+	ovi.scale_type = encoder->scale_type;
+
+	ovi.gpu_conversion = true;
+
+	mix = obs_create_video_mix(&ovi);
+	if (!mix)
+		return;
+
+	mix->encoder_only_mix = true;
+	mix->encoder_refs = 1;
+	mix->view = &obs->data.main_view;
+
+	pthread_mutex_lock(&obs->video.mixes_mutex);
+
+	// double check that nobody else added a matching mix while we've created our mix
```

Comment:
Why is a second check necessary here? Couldn't we hold the `mixes_mutex` while we create and add the mix?

Reply:
Not sure how expensive the individual parts inside `obs_create_video_mix` are, but it creates a few GPU resources and holds `mixes_mutex` for a while which would likely result in a deadlock if `mixes_mutex` is not a recursive mutex

I considered just creating the mix right away and only then checking whether it was necessary and adding it to the mixes list if needed, but traversing existing mixes seemed less expensive than unconditionally creating a new mix

I can double check whether holding the mutex while creating the new mix work though next week

Reply:
@notr1ch: I've added a few profiler sections in <https://github.com/kc5nra/obs-studio/compare/ruwen/core-mix-gpu-scaling-standalone...kc5nra:obs-studio:ruwen/core-mix-gpu-scaling-standalone-profile> which gave the following numbers:

- In a debug build:
```
14:20:58.758: == Profiler Results =============================
14:20:58.758: run_program_init: 4536.06 ms
14:20:58.759:  ┣OBSApp::AppInit: 59.214 ms
14:20:58.759:  ┃ ┗OBSApp::InitLocale: 3.394 ms
14:20:58.759:  ┗OBSApp::OBSInit: 4417.01 ms
14:20:58.759:    ┣obs_startup: 17.556 ms
14:20:58.759:    ┗OBSBasic::OBSInit: 3569.51 ms
14:20:58.759:      ┣OBSBasic::InitBasicConfig: 0.986 ms
14:20:58.759:      ┣OBSBasic::ResetAudio: 0.882 ms
14:20:58.759:      ┣OBSBasic::ResetVideo: 1916.84 ms
14:20:58.759:      ┃ ┗obs_create_video_mix: 7.367 ms
14:20:58.759:      ┃   ┣copy_fps: 0 ms
14:20:58.760:      ┃   ┣open_output: 4.024 ms
14:20:58.760:      ┃   ┣lock_gpu: 0.002 ms
14:20:58.760:      ┃   ┗create_gpu_resources: 3.273 ms
[...snip...]
14:20:58.766: maybe_set_gpu_video_media: min=0 ms, median=0 ms, max=4.959 ms, 99th percentile=4.959 ms
14:20:58.766:  ┣find_mix_pre_create: 0.001 ms, 0.5 calls per parent call
14:20:58.766:  ┣create_mix: 4.952 ms, 0.5 calls per parent call
14:20:58.766:  ┃ ┗obs_create_video_mix: 4.951 ms
14:20:58.766:  ┃   ┣copy_fps: 0 ms
14:20:58.766:  ┃   ┣open_output: 1.848 ms
14:20:58.766:  ┃   ┣lock_gpu: 0.001 ms
14:20:58.767:  ┃   ┗create_gpu_resources: 3.088 ms
14:20:58.767:  ┗find_mix_post_create: 0.001 ms, 0.5 calls per parent call
```

- in a release (`RelWithDebInfo`) build:
```
14:28:55.499: == Profiler Results =============================
14:28:55.499: run_program_init: 4084.23 ms
14:28:55.499:  ┣OBSApp::AppInit: 51.2 ms
14:28:55.500:  ┃ ┗OBSApp::InitLocale: 1.741 ms
14:28:55.500:  ┗OBSApp::OBSInit: 3980.21 ms
14:28:55.500:    ┣obs_startup: 13.025 ms
14:28:55.500:    ┗OBSBasic::OBSInit: 3543.19 ms
14:28:55.500:      ┣OBSBasic::InitBasicConfig: 1.036 ms
14:28:55.500:      ┣OBSBasic::ResetAudio: 0.608 ms
14:28:55.500:      ┣OBSBasic::ResetVideo: 2099.66 ms
14:28:55.500:      ┃ ┗obs_create_video_mix: 3.422 ms
14:28:55.500:      ┃   ┣copy_fps: 0 ms
14:28:55.500:      ┃   ┣open_output: 0.68 ms
14:28:55.500:      ┃   ┣lock_gpu: 0 ms
14:28:55.500:      ┃   ┗create_gpu_resources: 2.721 ms
[...snip...]
14:28:55.505: maybe_set_gpu_video_media: min=0 ms, median=0 ms, max=2.887 ms, 99th percentile=2.887 ms
14:28:55.505:  ┣find_mix_pre_create: 0 ms, 0.5 calls per parent call
14:28:55.505:  ┣create_mix: 2.885 ms, 0.5 calls per parent call
14:28:55.505:  ┃ ┗obs_create_video_mix: 2.885 ms
14:28:55.505:  ┃   ┣copy_fps: 0 ms
14:28:55.505:  ┃   ┣open_output: 0.666 ms
14:28:55.505:  ┃   ┣lock_gpu: 0 ms
14:28:55.505:  ┃   ┗create_gpu_resources: 2.209 ms
14:28:55.505:  ┗find_mix_post_create: 0 ms, 0.5 calls per parent call
```

I'm leaning towards leaving this as they are right now based on those numbers, but I'm open to changing it if there's consensus on whether holding a lock while creating the mix (which would require factoring out the relevant lock calls inside `obs_create_video_mix`) or just creating a new mix unconditionally is the approach to take

Reply:
Thanks for benchmarking, those numbers might indeed pose a risk to a 60 FPS render target. Though in practice, when would a video mix be created during an active stream / recording session? Wouldn't they be created ahead of time or on startup where this isn't a concern?

Reply:
They're created whenever new encoders are created (with output requirements that diverge from whatever is already active), so e.g. changing recording settings while the stream is active or changing stream settings while recording is active?
To be fair, the `create_gpu_resources` section might still pose a risk to rendering or similar since the graphics lock is being held for that duration